internal/service/advancedcluster/resource_update_logic.go

```diff
@@ -0,0 +1,61 @@
+package advancedcluster
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/mongodb/terraform-provider-mongodbatlas/internal/common/conversion"
+	"go.mongodb.org/atlas-sdk/v20240530002/admin"
+)
+
+func noIDsPopulatedInReplicationSpecs(replicationSpecs *[]admin.ReplicationSpec20250101) bool {
+	if replicationSpecs == nil || len(*replicationSpecs) == 0 {
+		return false
+	}
+	for _, spec := range *replicationSpecs {
+		if conversion.IsStringPresent(spec.Id) {
+			return false
+		}
+	}
+	return true
+}
+
+func populateIDValuesUsingNewAPI(ctx context.Context, projectID, clusterName string, connV2CLusterAPI admin.ClustersApi, replicationSpecs *[]admin.ReplicationSpec20250101) (*[]admin.ReplicationSpec20250101, diag.Diagnostics) {
+	if replicationSpecs == nil || len(*replicationSpecs) == 0 {
+		return replicationSpecs, nil
+	}
+	cluster, _, err := connV2CLusterAPI.GetCluster(ctx, projectID, clusterName).Execute()
+	if err != nil {
+		return nil, diag.FromErr(fmt.Errorf(errorRead, clusterName, err))
+	}
+
+	zoneToReplicationSpecsIDs := groupIDsByZone(cluster.GetReplicationSpecs())
+	result := AddIDsToReplicationSpecs(*replicationSpecs, zoneToReplicationSpecsIDs)
+	return &result, nil
+}
+
+func AddIDsToReplicationSpecs(replicationSpecs []admin.ReplicationSpec20250101, zoneToReplicationSpecsIDs map[string][]string) []admin.ReplicationSpec20250101 {
+	for zoneName, availableIDs := range zoneToReplicationSpecsIDs {
+		var indexOfIDToUse = 0
+		for i := range replicationSpecs {
+			if indexOfIDToUse >= len(availableIDs) {
+				break // all available ids for this zone have been used
+			}
+			if replicationSpecs[i].GetZoneName() == zoneName {
```

Comment:
so essentially we are setting an "id" for the patch request specs based on the zone name irrespective of whether the spec exists or not. And we don't care about setting the id on the "correct" matching spec as well.
In case new specs are added, some specs won't have an id for the same zone name.
Is this correct?

Reply:
> we are setting an "id" for the patch request specs based on the zone name

Correct, keep in mind that this transition is going from the old shard schema (either sharded or geo-sharded) which can only define symmetric shards for a zone. This is why we can hold this statement: `we don't care about setting the id on the "correct" matching spec as well`. If new specs are added they would not have an id, which is accurate in terms of the update request.